repo/content/index_blob_manager.go

```diff
@@ -0,0 +1,325 @@
+package content
+
+import (
+	"bytes"
+	"context"
+	"crypto/aes"
+	"encoding/hex"
+	"encoding/json"
+	"strings"
+	"time"
+
+	"github.com/pkg/errors"
+
+	"github.com/kopia/kopia/internal/gather"
+	"github.com/kopia/kopia/repo/blob"
+	"github.com/kopia/kopia/repo/encryption"
+	"github.com/kopia/kopia/repo/hashing"
+)
+
+// indexBlobManager is the API of index blob manager as used by content manager.
+type indexBlobManager interface {
+	writeIndexBlob(ctx context.Context, data []byte) (blob.Metadata, error)
+	listIndexBlobs(ctx context.Context, includeInactive bool) ([]IndexBlobInfo, error)
+	getIndexBlob(ctx context.Context, blobID blob.ID) ([]byte, error)
+	registerCompaction(ctx context.Context, inputs, outputs []blob.Metadata) error
+	flushCache()
+}
+type indexBlobManagerImpl struct {
+	st                            blob.Storage
+	hasher                        hashing.HashFunc
+	encryptor                     encryption.Encryptor
+	listCache                     *listCache
+	ownWritesCache                ownWritesCache
+	timeNow                       func() time.Time
+	indexBlobCache                contentCache
+	minIndexBlobDeleteAge         time.Duration
+	minCompactionLogBlobDeleteAge time.Duration
+}
+
+func (m *indexBlobManagerImpl) listIndexBlobs(ctx context.Context, includeInactive bool) ([]IndexBlobInfo, error) {
+	compactionLogMetadata, err := m.listCache.listIndexBlobs(ctx, compactionLogBlobPrefix)
+	if err != nil {
+		return nil, errors.Wrap(err, "error listing compaction log entries")
+	}
+
+	compactionLogMetadata, err = m.ownWritesCache.merge(ctx, compactionLogBlobPrefix, compactionLogMetadata)
+	if err != nil {
+		return nil, errors.Wrap(err, "error merging local writes for compaction log entries")
+	}
+
+	storageIndexBlobs, err := m.listCache.listIndexBlobs(ctx, indexBlobPrefix)
```

Comment:
Making two calls to listIndexBlobs with different prefixes one after the other is an issue here, consider this -
1. As part of some process X, the first call to list compaction logs gave us the blobs existing.
2. The process X then stalled for some reason for a long time (hours) before listing all index blobs.
3. Another process Y came in and wrote a compaction log blob, compacting some index blobs say A and B.
4. After some time, another process Z, removed index blobs A and B.
5. Just after process Z removed A, process X resumed again to now read index blob B. This is an issue as it did not read the earlier compaction blob that was written.

Maybe we should first list all index blobs, and then all compaction log blobs.